Subscriber chart improvements

guarani opened the pull request:
Fixes https://github.com/wordpress-mobile/WordPress-iOS/issues/23044

This PR addresses the remaining tasks from https://github.com/wordpress-mobile/WordPress-iOS/pull/23074.

# Changes

## Make it easier to see changes in subscriber growth

The line chart often looked flat, and small fluctuations in subscribers were not visible. I updated the y-axis minimum & maximum to cover just the range of values shown on the chart, just like on Android (see https://github.com/wordpress-mobile/WordPress-iOS/pull/23074#issuecomment-2077130702).

## Add markers to the chart

I added marker support to match the designs. Tapping any point on the chart shows the subscriber count on that day in a tooltip.

# To test

## Visualizing change

Test a site with no subscriber change in the last month, test one with some change, and test a site with lots of change. Here's how it looks:

| No change | Little change  | Lots of change |
| - | - | - |
| <img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1898325/766f6c77-efa3-473f-af9a-920e714ca64c" /> | <img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1898325/752955ac-3680-42f3-ad1f-e3ad39b72183" /> | <img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1898325/9b3b25d9-0739-4a39-8904-4d699ba17912" /> |

There are still some known issues:
- The chart may occasionally show spikes in subscriber growth (pull-to-refresh normally fixes it) due to a backend issue: pejTkB-1l5-p2#comment-1334
- The y-axis should not abbreviate labels when its range is small. In the "Lots of change" screenshot above, instead of all the labels being `1.08k`, the labels should be `1,080`, etc. like in the screenshot below
  <details>
  <summary>See WP.com screenshot</summary>
  <img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1898325/194a8f98-498f-4d03-b236-d5a8788f33e3" />
  </details>

## Chart markers

The chart now shows a marker when tapped. This works very similarly to the marker on the Views & Visitors chart, but matches the new designs:

<img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1898325/d763e795-c967-4cce-a2f0-4578c2760861" width="250" />

## Regression Notes
1. Potential unintended areas of impact

The Views & Visitor chart and its marker could be affected by these changes, though care was taken not to.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Manually testing the Views & Visitors chart for regressions.

3. What automated tests I added (or what prevented me from doing so)

I didn't add unit tests because most of the logic is for the chart (a UI component) and most of the code is chart configuration, not logic

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

Commit:
Zoom subscriber chart to values

Commit:
Add Subscriber chart marker

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

Commit:
Split StatsChartMarker into separate files

Split into a base class, a SubscribersChartMarker class, and a ViewsVisitorsChartMarker class

Commit:
Fix views & visitors y-axis min regression

guarani commented on WordPress/Classes/ViewRelated/Stats/Charts/StatsLineChartView.swift:
```diff
@@ -246,22 +254,39 @@ private extension StatsLineChartView {
+                }
+            } else if statType == .subscribers {
+                yAxis.axisMinimum = data.getYMin(axis: .left)
+                yAxis.axisMaximum = data.getYMax(axis: .left)
```
I didn't use `VerticalAxisFormatter.roundUpAxisMaximum()` here because it results in y-axis maximums that are too big. However, the downside to not using it is that the y-axis values aren't "neat" round numbers e.g. "1.57k" might be used instead of "1.8k" as one of the y-axis tick labels.

Commit:
Clean up StatsChartMarker.text(for:)

guarani commented on WordPress/Classes/ViewRelated/Stats/Subscribers/SubscribersChartMarker.swift:
```diff
@@ -0,0 +1,35 @@
+                                                                          .paragraphStyle: paragraphStyle,
+                                                                          .foregroundColor: UIColor.white.withAlphaComponent(0.8)]
+
+        let attrString = NSMutableAttributedString()
+        attrString.append(NSAttributedString(string: entry.y.abbreviatedString() + "\n", attributes: subscriberCountRowAttributes))
+        attrString.append(NSAttributedString(string: "\(name)\n", attributes: subscriberLabelRowAttributes))
+        attrString.append(NSAttributedString(string: DateValueFormatter().dateFormatter.string(from: date), attributes: subscriberDateRowAttributes))
+        return attrString
```
The issue with this approach is that "8 Views" isn't reversed for languages where the number follows the noun. It's an existing issue with the markers on the Views & Visitors card but it's repeated here.

staskus replied:
Good observation.

When it's written vertically I wonder if it should be read like a sentence or more like a title and subtitle. Maybe then the language order doesn't matter as much?

```
8
Subscribers
Apr 11
```

guarani replied:
I'm not sure. I'm ok with leaving it out-of-scope of this PR because it's not a straightforward fix and as you say, it might not be such a big issue after all. I'll create a separate issue to cover both the Subscriber chart markers and the Views & Visitor chart markers, so we can decide there what to do.

Commit:
Fix duplicate tick marks

staskus commented on WordPress/Classes/ViewRelated/Stats/Subscribers/SubscribersChartMarker.swift:
```diff
@@ -0,0 +1,35 @@
+                                                                          .paragraphStyle: paragraphStyle,
+                                                                          .foregroundColor: UIColor.white.withAlphaComponent(0.8)]
+
+        let attrString = NSMutableAttributedString()
```
We could make the line spacing a bit smaller to align more closely with designs:

![image](https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/7e6468f2-1aad-4dd4-945b-c19b424a4fe1)

guarani replied:
You're right, seven ticks are shown on the designs 👍 Thanks for fixing this!

staskus commented on WordPress/Classes/ViewRelated/Stats/Charts/StatsLineChartView.swift:
```diff
@@ -232,36 +241,62 @@ private extension StatsLineChartView {
+                leftAxis.axisMaximum = lowestMaxValue
+            }
+        } else if statType == .subscribers {
+            if areDataValuesIdentical {
```
Maybe we could simplify it by checking this value within the chart itself.

```suggestion
            if data.getYMax(axis: .left) == data.getYMin(axis: .left) {
```

guarani replied:
This might have been an oversight. Thanks for spotting it!

guarani replied:
Fixed in 0924f99ae5f8aea34cf71f13f6491699fcd10b07! Makes the code much simpler 😅

staskus reviewed (commented):
Thanks! 

It works as expected. I made couple of code-related, and couple of UI-related ones. 

### A couple of UI observations

1. The chart line becomes narrower when it's on the top or the bottom axis, which is quite visible with this configuration where we set chart y axis minimum and maximum to be equal to the data set's y minimum and maximum

![image](https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/6be925ec-06b7-44e4-a32c-b4b0c1624e8d)

2. Too much gradation of x-axis compared to designs

![image](https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/d222ebdb-e0eb-4898-92e3-56340c9d3ed0)

staskus committed:
Limit maximum number of xAxis ticks

staskus committed:
Make line spacing smaller between title and number in chart marker

staskus committed:
Make chart line appear full width on top or bottom axis

staskus committed:
Only add filter dimension property to analytics for viewsAndVisitors type line chart

staskus commented:
I made a couple of improvements I thought were needed:

- Limit number of x-axis ticks https://github.com/wordpress-mobile/WordPress-iOS/pull/23119/commits/d543db3072a4a10f24ddd4b0ab266b0fc5313153
- Make line spacing smaller in chart marker https://github.com/wordpress-mobile/WordPress-iOS/pull/23119/commits/f5220905d09578af02887b754a09305611499990
- Make line appear full width when it's on top or at the bottom of axis https://github.com/wordpress-mobile/WordPress-iOS/pull/23119/commits/0e720b3d4594751289e4836322d058af15630060
- Fix subscribers line tap analytics being tracked with "views" property https://github.com/wordpress-mobile/WordPress-iOS/pull/23119/commits/e2a1ff7f7d73eddd21a2e29cf9eb4e15e83a3142

| Before | After |
|--------|-------|
| ![image](https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/a5d1b00a-0c87-4b46-91b3-7c3e565167e2) | ![After-Subscribers](https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/5772d6c6-259e-4cf2-93df-31fa15fa25be) |

Commit:
Simplified areDataValuesIdentical

guarani commented:
> * Fix subscribers line tap analytics being tracked with "views" property [e2a1ff7](https://github.com/wordpress-mobile/WordPress-iOS/commit/e2a1ff7f7d73eddd21a2e29cf9eb4e15e83a3142)

Good catch!

I just checked which Tracks events are currently fired when tapping points on the chart and saw:
```
 Tracked: stats_line_chart_tapped <granularity: days>
```

Doing the same on the Views & Visitors chart shows:
```
🔵 Tracked: stats_line_chart_tapped <granularity: days, property: views>
```
While I could add a `property: subscribers` here, I'll leave that for https://github.com/wordpress-mobile/WordPress-iOS/issues/23058 since we haven't discussed Tracks events yet for this AFAIK.

guarani commented:
This is ready for a re-review @staskus. Thanks for all the help moving this forward!
The only change I made was to simplify the logic around `areDataValuesIdentical`.
